neo4j-to-neptune/src/main/java/com/amazonaws/services/neptune/util/NeptuneBulkLoader.java

```diff
@@ -270,66 +273,117 @@ private CompletableFuture<Boolean> uploadFilesSequentially(File[] files, String
-                            completedUpload.response().eTag());
+            System.err.println("Initiating Transfer Manager upload with compression...");
+            Upload upload = transferManager.upload(uploadRequest);
+
+            // Wait for BOTH upload and compression to complete - fail if either fails
+            return CompletableFuture.allOf(upload.completionFuture(), compressionFuture)
+                .thenApply(ignored -> {
+                    System.err.println(
+                        "Successfully uploaded " + localFile.getName() +
+                        " (compressed) - ETag: " + upload.completionFuture().join().response().eTag());
                     return true;
+                })
+                .exceptionally(throwable -> {
+                    logUploadError(localFilePath, throwable);
+                    // Re-throw to maintain fail-fast behavior
+                    if (throwable instanceof RuntimeException) {
+                        throw (RuntimeException) throwable;
+                    } else {
+                        throw new RuntimeException("Upload or compression failed", throwable);
+                    }
+                })
+                .whenComplete((result, throwable) -> {
+                    // Always shutdown the executor service
+                    streamExecutor.shutdown();
+                });
+        }
+    }
+
+    /**
+     * Start the compression task in a background thread
+     */
+    protected CompletableFuture<Void> startCompressionTask(File localFile, PipedOutputStream pipedOut) {
+        return CompletableFuture.runAsync(() -> {
+            try (GZIPOutputStream gzipOut = new GZIPOutputStream(pipedOut);
+                 FileInputStream fis = new FileInputStream(localFile);
+                 BufferedInputStream fileIn = new BufferedInputStream(fis)) {
+
+                // Use same buffer size as GzipCompressUtils for consistency
+                byte[] buffer = new byte[8192];
+                int bytesRead;
+                while ((bytesRead = fileIn.read(buffer)) != -1) {
+                    gzipOut.write(buffer, 0, bytesRead);
                 }
-            });
+                gzipOut.finish();
 
-        } catch (Exception e) {
-            System.err.println("Error initiating Transfer Manager upload for " + localFilePath + ": " + e.getMessage());
-            e.printStackTrace();
-            return CompletableFuture.completedFuture(false);
+            } catch (IOException e) {
+                throw new UncheckedIOException("Compression failed for " + localFile.getAbsolutePath(), e);
+            }
+        });
+    }
+
+    /**
+     * Create the S3 upload request with compressed stream
+     */
+    private UploadRequest createUploadRequest(String s3Key, PipedInputStream pipedIn, ExecutorService streamExecutor) {
+        return UploadRequest.builder()
+            .putObjectRequest(PutObjectRequest.builder()
```

Comment:
Nit: can use consumer builder overload instead
```suggestion
            .putObjectRequest(putBuilder -> putBuilder
```

Reply:
I prefer the PutObjectRequest.builder() in this case